### PR TITLE
LinReg extended API

### PR DESCRIPTION
**Pull request by KulikovNikita**
**Major changes:**
1) Enables various result options for the `train` stage, including: `intercept`, `coefficients`
2) Eliminates unnecessary abstract class usage in Linear Regression model. Now it is implemented via simple pimpl to concrete class
3) Extends testing to ensure that `intercept`s and `coeffiicients` are computed correctly

**Minor changes:**
4) Extends debug functionality

### CODE REVIEW

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ static train_result<Task> call_dal_kernel(const context_gpu& ctx,
     using dal::detail::check_mul_overflow;
 
     using model_t = model<Task>;
-    using model_impl_t = backend::norm_eq_model_impl<Task>;
+    using model_impl_t = detail::model_impl<Task>;
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:61]: I do not fully understand why norm_eq_model_impl was changed to model_impl in certain places?
As I understand norm_eq_model_impl specifies that this model was produced using a normal equations method (norm_eq). In the future we might add other methods to solve linear regression problem (QR, gradient descent, ...).
For me, all the *_model_impl classes for various compute methods will have common parts: coefficiens and intercept; but might have other differences.
What I do not understand: is the current version of the code scales well in case of adding other types of *_model_impl?

File: cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp [hunk 2]
```diff
@@ -22,43 +22,33 @@
 
 namespace oneapi::dal::linear_regression {
 
+using norm_eq_proto = ONEDAL_SERIALIZABLE(linear_regression_model_impl_id);
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp:25]: The same question: why the ID linear_regression_norm_eq_model_impl_id was changed to linear_regression_model_impl_id?
If we would add a model for QR compute method, for example, that model might have different internal structure than norm_eq model, so it will need a different serialization ID.

File: cpp/oneapi/dal/algo/linear_regression/common.cpp [hunk 3]
```diff
@@ -60,14 +95,25 @@ namespace v1 {
...
+model<Task>::model(const std::shared_ptr<detail::model_impl<Task>>& impl) : impl_{ impl } {}
 
 template <typename Task>
 const table& model<Task>::get_betas() const {
-    return impl_->get_betas();
+    return impl_->get_packed_coefficients();
+}
+
+template <typename Task>
+const table& model<Task>::get_packed_coefficients() const {
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/common.cpp:109]: Should we also extend the docs as new API methods were added?
BTW, I do not see the docs for linear regression at all in github.io.

File: cpp/oneapi/dal/algo/linear_regression/train_types.cpp [hunk 4]
```diff
@@ -33,6 +34,11 @@ class detail::v1::train_input_impl : public base {
 template <typename Task>
 class detail::v1::train_result_impl : public base {
 public:
+    table intercept;
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/train_types.cpp:37]: I see that it is possible to get an inconsistent state of train_result_impl when storing intercept and coefficients tables and model as a separate fields of the result. As they might be modified independently from each other.
Do we need set_intercept() and set_coefficients() at all? They only can lead to an object inconsistency and are not needed for compatibility with sklearn.

File: cpp/oneapi/dal/algo/linear_regression/train_types.hpp [hunk 5]
```diff
@@ -98,9 +98,47 @@ class train_result {
         return *this;
     }
 
+    /// Table of Linear regression intercept
+    const table& get_intercept() const;
+
+    auto& set_intercept(const table& value) {
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/train_types.hpp:104]: Do we need this?
The same question for set_coefficients, set_packed_coefficients.

File: cpp/oneapi/dal/backend/serialization.hpp [hunk 6]
```diff
@@ -187,7 +187,7 @@ class serialization_ids {
     ID(6040000000, decision_forest_model_interop_impl_reg_id);
 
     // Algorithms - Linear Regression
-    ID(7010000000, linear_regression_norm_eq_model_impl_id);
+    ID(7010000000, linear_regression_model_impl_id);
```
Comment (Vika-F) [inline cpp/oneapi/dal/backend/serialization.hpp:190]: See the question about serialization ID above.

**Review by Vika-F — COMMENTED** [review]
Need some clarifications before I can approve this PR.

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp [hunk 7]
```diff
@@ -109,7 +109,7 @@ static infer_result<Task> call_dal_kernel(const context_gpu& ctx,
     ONEDAL_ASSERT((f_count + 1) == betas.get_column_count());
 
     const auto resps_size = check_mul_overflow(s_count, r_count);
-    auto resps_arr = array<Float>::empty(queue, resps_size);
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp:112]: I suggest not to mix enhancement and bug fixing in one PR

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp [hunk 8]
```diff
@@ -109,7 +109,7 @@ static infer_result<Task> call_dal_kernel(const context_gpu& ctx,
...
+    auto resps_arr = array<Float>::zeros(queue, resps_size, alloc);
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp:112]: I suggest not to mix enhancement and bug fixing in one PR

File: cpp/oneapi/dal/algo/linear_regression/common.cpp (hunk 3 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/common.cpp:109]: Just because there are no such docs for all new OneAPI algorithms at all. And yes it is in todo list but not with a low priority

File: cpp/oneapi/dal/algo/linear_regression/backend/cpu/train_kernel_norm_eq.cpp [hunk 9]
```diff
@@ -51,30 +52,27 @@ static train_result<Task> call_daal_kernel(const context_cpu& ctx,
     using dal::detail::check_mul_overflow;
 
     using model_t = model<Task>;
-    using model_impl_t = backend::norm_eq_model_impl<Task>;
+    using model_impl_t = detail::model_impl<Task>;
 
     bool intp = desc.get_compute_intercept();
 
-    const auto feature_count = data.get_column_count();
-    const auto response_count = resp.get_column_count();
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/cpu/train_kernel_norm_eq.cpp:59]: 1) It is just too wordy
2) Naming should be consistent between CPU & GPU parts

But okay I'll fix it

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp (hunk 1 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:61]: As you have already said there is no many options what can be inside of the model. Only coefficients and possibly intercept values. So there is no way that another implementation of model should be added. In any way it can be restored relatively easy.

Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:61]: And btw you can just extend this model with empty tables if needed. Anyway the method of computations is determined based on descriptor, not on runtime info.

File: cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp (hunk 2 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp:25]: Just because all models can be converted to the form: intercept + coefficients

File: cpp/oneapi/dal/algo/linear_regression/train_types.cpp (hunk 4 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/train_types.cpp:37]: Yes. In some sense these tables are not necessarily required since they can be obtained by decomposition of `packed_coefficients` table.
The thing is:
1) `train_results` is a separate entity from a `model`. We do not care about this tables after them being set. The infer stage is fed only with `model` itself
2) It can be hard to split table that is stored inside the model. Our table API does not support any kind of table views
3) User algorithms should be more or less agnostic about the internal structure of the model. In example there are many algorithms like kd-tree or decision-forest when user do not have access to the guts of model at all.

File: cpp/oneapi/dal/backend/serialization.hpp (hunk 6 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/serialization.hpp:190]: We need to save this ID for the sake of backward compatibility between models. Naming is changed according to a new class name

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp (hunk 7 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/infer_kernel_norm_eq_dpc.cpp:112]: Agreed. But in this case these fixes are pretty small

File: cpp/oneapi/dal/backend/serialization.hpp (hunk 6 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/serialization.hpp:190]: Discussed

File: cpp/oneapi/dal/algo/linear_regression/train_types.hpp (hunk 5 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/train_types.hpp:104]: DIscussed

File: cpp/oneapi/dal/algo/linear_regression/train_types.cpp (hunk 4 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/train_types.cpp:37]: Discussed

File: cpp/oneapi/dal/algo/linear_regression/common.cpp (hunk 3 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/common.cpp:109]: Discussed

File: cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp (hunk 2 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp:25]: Discussed

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp (hunk 1 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:61]: Discussed offline

File: cpp/oneapi/dal/algo/linear_regression/backend/cpu/train_kernel_norm_eq.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/cpu/train_kernel_norm_eq.cpp]: Fixed both in oneapi and daal parts

**Review by Vika-F — APPROVED** [review]
The changes look good to me.

File: cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp [hunk 10]
```diff
@@ -22,43 +22,33 @@
...
+
 template <typename Task>
-class detail::v1::model_impl : public base {
+class detail::v1::model_impl : public norm_eq_proto {
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/linear_regression/backend/model_impl.hpp:28]: I would call it like `norm_eq_base`, but that's fine too.

File: cpp/oneapi/dal/algo/linear_regression/test/batch.cpp [hunk 11]
```diff
@@ -34,10 +34,10 @@ class lr_batch_test : public lr_test<TestType, lr_batch_test<TestType>> {
...
 
     void generate_dimensions() {
-        this->t_count_ = GENERATE(327, 888, 32777);
-        this->s_count_ = GENERATE(111, 333, 16777);
-        this->f_count_ = GENERATE(2, 17, 99);
-        this->r_count_ = GENERATE(2, 15, 33);
+        this->t_count_ = GENERATE(307, 12999, 17777);
+        this->s_count_ = GENERATE(111, 777);
+        this->f_count_ = GENERATE(2, 17);
+        this->r_count_ = GENERATE(2, 15);
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/linear_regression/test/batch.cpp:40]: Reducing dims?

**Review by samir-nasibli — APPROVED** [review]
LGTM

File: cpp/oneapi/dal/algo/linear_regression/test/batch.cpp (hunk 11 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/test/batch.cpp:40]: Yeah. It takes just too much time to evaluate all variants especially on relatively weak hardware
